_source/_posts/2021-10-08-private-key-jwt-python.md

```diff
@@ -0,0 +1,167 @@
+
+```
+PRIVATE_KEY=
+PUBLIC_KEY=
+MODULUS=
+CLIENT_ID=
+OKTA_URL=
+API_KEY=
+SCOPES=
+```
+
+Scripts will generate values for the first 4 vars. You will need to fill out the other values. **Note:** the API key is needed for creating the OAuth client in Okta, so make sure the token you use has permissions to create applications.
+
+- OKTA_URL=https://{your_okta_domain}
+- API_KEY={okta_api_key}
+- SCOPES='okta.users.read okta.users.manage'
+
+*The SCOPES var is an string of scopes (seperated by spaces) granted to your eventual access token that you will be using for management tasks. I put in a couple common ones as a placeholder. In your Okta application, it would look like this:*
+
+![SCOPES_1](https://i.imgur.com/B4zNpom.png)
+
+## Running the scripts
+
+To run these scripts, make sure you are still in the root folder of the cloned repository.
+
+### Step 1: Generate keypair
+
+Run the following command in terminal:
+
+```
+python generate_keys.py
+```
+
+You should see confirmation printed to the terminal:
+
+```
+///////////////// Keys generated and added to .env file.
+```
+
+Sure enough, you can pop over to the .env file to see the `PRIVATE_KEY`, `PUBLIC_KEY` and `MODULUS` values populated there.
+
+### Step 2: Create OAuth service app in Okta
+
+Now that you have the keypair, you can create the OAuth app in Okta with the public key. The scopes you defined in the .env file will also be granted in the application. Type the following into the terminal:
+
+```
+python create_okta_service_app.py
+```
+
+You will get confirmation in the terminal of the created app, that the client ID of the app was updated in your .env file, as well as scopes granted.
+
+```
+///////////////// Okta Service app created: 
+
+{application object}
+
+///////////////// Setting CLIENT_ID in .env:
+
+{client ID of newly created app}
+
+///////////////// Scopes granted: 
+
+{scopes granted to the application}
+```
+
+If you navigate to your applications in Okta, you will see this newly created app under the name **Private Key JWT Service App**.
+
+### Step 3: Create PKJ and get access token
+
+The two scripts above should only need to run once for setup. From hereon out, you can run the following command to get an access token from Okta:
+
+```
+python get_access_token.py
+```
+
+You should see the access token printed to terminal:
+
+```
+///////////////// Access Token:
+
+{access_token}
+```
+
+You can now use this token to perform management tasks on your Okta org based on your defined scopes by passing it in the **Authorization** header as **Bearer {access_token}**.
+
+## In conclusion: Keep it secret, keep it safe
+
+The Private Key JWT flow is complicated to set up, but once the pieces are in place it should be pretty automated. As always, the private key is the key to the castle. So...
+
+![Keep it secret, keep it safe!](https://thumbs.gfycat.com/HonestSmartBuffalo-max-1mb.gif)
```

Comment:
The animated gif linked here is broken in preview mode. Not sure how to verify or resolve.